### ISSUE TITLE
RESP3 PUSH messages incorrectly use pending callback

### ISSUE THREAD

**Issue by lukepalmer** [redis/hiredis#825]
I think that RESP3 PUSH data can be incorrectly sent to a callback that is waiting for a reply.

async.c makes the following assumption:
```
        /* Even if the context is subscribed, pending regular callbacks will                                                                                                       
         * get a reply before pub/sub messages arrive. */                                                                                                                          
        if (__redisShiftCallback(&ac->replies,&cb) != REDIS_OK) {                                                                                                                  

```
I think this is not true because the RESP3 spec says:

Push data may be interleaved with any protocol data, but always at the top level, so the client will never find push data in the middle of a Map reply for instance.

Clients should normally react to the publication of a push data by invoking a callback associated with the push data type.

**Comment by michael-grunder** [comment]
Yes, at present hiredis can parse `>` push messages but there's no logic to handle those kinds of "out of band" messages properly.

This would help redis-cli too:
antirez/redis/issues/7125

@yossigo I kind of think we should support this before releasing v1.0.0.  I don't think it will be all that involved.

**Comment by michael-grunder** [comment]
I put together a draft for the proper handling of RESP3 `PUSH` messages.

Basically the commit allows users to specify a custom handler for `PUSH` messages if they wish in sync or async hiredis.

In order to maintain backward compatibility, the async API always handles non-subscribe `PUSH` messages out of band but if no callback is supplied by the user all we do is call `freeReplyObject` and continue the `redisGetReply` loop.

I'll do some more robust testing but the branch seems to work pretty well.

**Comment by michael-grunder** [comment]
Closed via #841
